frame/staking/src/voter_bags.rs

```diff
@@ -0,0 +1,883 @@
+		let weight_of = Pallet::<T>::weight_of_fn();
+
+		Self::insert_many(
+			nominators_iter.chain(validators_iter),
+			weight_of,
+		)
+	}
+
+	/// Decode the length of the voter list.
+	pub fn decode_len() -> Option<usize> {
+		crate::VoterCount::<T>::try_get().ok().map(|n| n.saturated_into())
+	}
+
+	/// Iterate over all nodes in all bags in the voter list.
+	///
+	/// Full iteration can be expensive; it's recommended to limit the number of items with `.take(n)`.
+	pub fn iter() -> impl Iterator<Item = Node<T>> {
+		// We need a touch of special handling here: because we permit `T::VoterBagThresholds` to
+		// omit the final bound, we need to ensure that we explicitly include that threshold in the
+		// list.
+		//
+		// It's important to retain the ability to omit the final bound because it makes tests much
+		// easier; they can just configure `type VoterBagThresholds = ()`.
+		let thresholds = T::VoterBagThresholds::get();
+		let iter = thresholds.iter().copied();
+		let iter: Box<dyn Iterator<Item = u64>> = if thresholds.last() == Some(&VoteWeight::MAX) {
+			// in the event that they included it, we can just pass the iterator through unchanged.
+			Box::new(iter.rev())
+		} else {
+			// otherwise, insert it here.
+			Box::new(iter.chain(iter::once(VoteWeight::MAX)).rev())
+		};
+		iter
+			.filter_map(Bag::get)
+			.flat_map(|bag| bag.iter())
+	}
+
+	/// Insert a new voter into the appropriate bag in the voter list.
+	///
+	/// If the voter is already present in the list, their type will be updated.
+	/// That case is cheaper than inserting a new voter.
+	pub fn insert_as(account_id: &AccountIdOf<T>, voter_type: VoterType) {
+		// if this is an update operation we can complete this easily and cheaply
+		if !Node::<T>::update_voter_type_for(account_id, voter_type) {
+			// otherwise, we need to insert from scratch
+			let weight_of = Pallet::<T>::weight_of_fn();
+			let voter = Voter { id: account_id.clone(), voter_type };
+			Self::insert(voter, weight_of);
+		}
+	}
+
+	/// Insert a new voter into the appropriate bag in the voter list.
+	pub fn insert(voter: VoterOf<T>, weight_of: impl Fn(&T::AccountId) -> VoteWeight) {
+		Self::insert_many(sp_std::iter::once(voter), weight_of);
+	}
+
+	/// Insert several voters into the appropriate bags in the voter list.
+	///
+	/// This is more efficient than repeated calls to `Self::insert`.
+	pub fn insert_many(
+		voters: impl IntoIterator<Item = VoterOf<T>>,
+		weight_of: impl Fn(&T::AccountId) -> VoteWeight,
+	) -> u32 {
+		let mut bags = BTreeMap::new();
+		let mut count = 0;
+
+		for voter in voters.into_iter() {
+			let weight = weight_of(&voter.id);
+			let bag = notional_bag_for::<T>(weight);
+			bags.entry(bag).or_insert_with(|| Bag::<T>::get_or_make(bag)).insert(voter);
+			count += 1;
+		}
+
+		for (_, bag) in bags {
+			bag.put();
+		}
+
+		crate::VoterCount::<T>::mutate(|prev_count| *prev_count = prev_count.saturating_add(count));
```

Comment:
also worried about count here with repeat items

Reply:
At this level of abstraction, nothing prevents repeat items. It would be a logic error for the system as a whole if one voter appeared twice in the data structure, but as far as keeping `count` accurate, insertion is infallible.